### PR TITLE
Reorganize the project folders

### LINKED ISSUE #2080: Reorganize the project folders
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently the datafusion related modules are organized directly under the project. It's better to downgrade them and organize them under one directory. And after #2076, the data access layer no longer depends on the datafusion, it's better to make it directly under the project.

**Describe the solution you'd like**

Then the whole project folders will be organized like follows:
Screen Shot 2022-03-25 at 10 17 28 AM

**Describe alternatives you've considered**

**Additional context**

It's blocked by #2076.

### PR DESCRIPTION
**Pull request by yahoNanJing**
# Which issue does this PR close?

Closes #2080.

 # Rationale for this change

# What changes are included in this PR?

Screen Shot 2022-03-28 at 9 03 11 AM

# Are there any user-facing changes?

### CODE REVIEW

File: data-access/README.md [hunk 1]
```diff
@@ -17,8 +17,6 @@
   under the License.
 -->
 
-# DataFusion Storage
+# Data Access Layer
 
-This module contains an `async` API for the [DataFusion][df] to access data, either remotely or locally.
-
-[df]: https://crates.io/crates/datafusion
+This module contains an `async` API for accessing data based on object store interfaces, either remotely or locally.
```
Comment (matthewmturner) [inline data-access/README.md:22]: If i understand the intent of this correctly, i think the data access layer could expose other data access methods than just object store (although they may not exist yet). For example a streaming provider, or database provider.  If this is correct i think we would just want to make the description more generic.

**Comment by matthewmturner** [comment]
I think its very clean / organized and i do prefer this proposed structure.  But im curious why an intermediate rust directory is needed?  do you have an idea or plans for code that would fall outside of that?

**Comment by liukun4515** [comment]
Do we need to align with this https://github.com/apache/arrow-datafusion/issues/1750?

File: Cargo.toml [hunk 2]
```diff
@@ -17,21 +17,21 @@
...
-    "datafusion-jit",
-    "datafusion-physical-expr",
-    "datafusion-cli",
+    "datafusion/rust/core",
+    "datafusion/rust/common",
+    "datafusion/rust/expr",
+    "datafusion/rust/jit",
+    "datafusion/rust/physical-expr",
+    "datafusion/rust/cli",
+    "datafusion/rust/proto",
```
Comment (houqp) [inline Cargo.toml:26]: the rust directory here seems a bit redundant to me, do you expect us to have other language implementations here?

File: data-access/Cargo.toml [hunk 3]
```diff
@@ -16,8 +16,8 @@
 # under the License.
 
 [package]
-name = "datafusion-storage"
-description = "Storage for DataFusion query engine"
+name = "data-access"
```
Comment (houqp) [inline data-access/Cargo.toml:19]: `data-access` feel a bit too generic to me. To be more specific, how about we name this crate and the folder `object-store`? the object-store crate hasn't been claimed in crates.io yet.

File: data-access/Cargo.toml [hunk 4]
```diff
@@ -16,8 +16,8 @@
...
+description = "General data access layer based on object store"
```
Comment (houqp) [inline data-access/Cargo.toml:20]: :+1: This is certainly a more accurate description :D

File: data-access/Cargo.toml (hunk 3 above)
Comment (matthewmturner) [reply data-access/Cargo.toml:19]: i had the idea that within this other data access providers / crates could exist - for example maybe stream provider or database provider and thats why the name was generic.

File: data-access/README.md (hunk 1 above)
Comment (yahoNanJing) [reply data-access/README.md:22]: Agree. Although currently we only support object store, we can make description not based on the object store.

**Comment by yahoNanJing** [comment]
> I think its very clean / organized and i do prefer this proposed structure. But im curious why an intermediate rust directory is needed? do you have an idea or plans for code that would fall outside of that?

Thanks @matthewmturner, @houqp , I'm good with removing the rust folder. Previously I was thinking maybe there's something datafusion specified, like Ballista did.

File: Cargo.toml (hunk 2 above)
Comment (yahoNanJing) [reply Cargo.toml:26]: Agree

File: data-access/Cargo.toml (hunk 3 above)
Comment (yahoNanJing) [reply data-access/Cargo.toml:19]: As @matthewmturner mentioned, the object store may be too specific and maybe will limit the extension of future data access abilities.

**Comment by yahoNanJing** [comment]
> Do we need to align with this https://github.com/apache/arrow-datafusion/issues/1750?

Yes, it's aligned with #1750.

**Comment by alamb** [comment]
I think we need to be careful if we make a crate with a name like `data-access` as to publish a new version of `datafusion` to crates.io we will also need to publish a version of `data-access`. Perhaps we can call it `datafusion-data-access` if we decide to go down this route

**Comment by yahoNanJing** [comment]
> I think we need to be careful if we make a crate with a name like `data-access` as to publish a new version of `datafusion` to crates.io we will also need to publish a version of `data-access`. Perhaps we can call it `datafusion-data-access` if we decide to go down this route

Agree. Will change "data-access" to be "datafusion-data-access".

**Comment by yahoNanJing** [comment]
Rebase and add refinement according to the PR review comments.

**Review by houqp — APPROVED** [review]
LGTM, the next datafusion release is going to be fun since we will have a lot of versions to update :D

**Review by yjshen — APPROVED** [review]

**Comment by yahoNanJing** [comment]
Hi @alamb, any chance to merge this PR this week😂

**Comment by alamb** [comment]
@yahoNanJing  sounds good to me! I haven't been following it closely but given @yjshen and @houqp  have given it ✅  I'll plan to merge it in when the tests pass

Thanks for sticking with it

**Comment by alamb** [comment]
Interestingly the failure in ballista also fails for me locally: https://github.com/apache/arrow-datafusion/runs/5726864404?check_suite_focus=true

Maybe something about moving the code into different directories has caused the test to start running

fixed in463de7f97

**Comment by yahoNanJing** [comment]
Thanks @alamb for fixing the issue.

**Comment by yjshen** [comment]
Thanks again @yahoNanJing and everyone else helped!

**Pull request merged**